dipy/io/stateful_tractogram.py

```diff
@@ -184,16 +213,77 @@ def __str__(self):
         text += '\nstreamline_count: {}'.format(self._get_streamline_count())
         text += '\npoint_count: {}'.format(self._get_point_count())
         text += '\ndata_per_streamline keys: {}'.format(
-            self.data_per_point.keys())
+            self.get_data_per_streamline_keys())
         text += '\ndata_per_point keys: {}'.format(
-            self.data_per_streamline.keys())
+            self.get_data_per_point_keys())
 
         return text
 
     def __len__(self):
         """ Define the length of the object """
         return self._get_streamline_count()
 
+    def __getitem__(self, key):
+        """ Slice all data in a consistent way """
+        if isinstance(key, int):
+            key = [key]
+
+        return self.from_sft(self.streamlines[key], self,
+                             data_per_point=self.data_per_point[key],
+                             data_per_streamline=self.data_per_streamline[key])
+
+    def __eq__(self, other):
+        """ Robust StatefulTractogram equality test """
+        if not self.is_compatible_sft(self, other):
+            return False
+
+        streamlines_equal = np.allclose(self.streamlines.get_data(),
+                                        other.streamlines.get_data())
+        dpp_equal = True
+        for key in self.data_per_point:
+            dpp_equal = dpp_equal and np.allclose(
+                self.data_per_point[key].get_data(),
+
+                other.data_per_point[key].get_data())
+        dps_equal = True
+        for key in self.data_per_streamline:
+            dps_equal = dps_equal and np.allclose(
+                self.data_per_streamline[key],
+                other.data_per_streamline[key])
+
+        return streamlines_equal and dpp_equal and dps_equal
```

Comment:
Nice! I like this! However, I wonder why should we continue if `streamlines_equal` is False. it avoids us to go through the 2 others loop.  Same comment if `dpp_equal` is False.

So I think, you skip or quit the function as soon as one test is False